deploy/stats.go

```diff
@@ -58,66 +291,154 @@ type RunStats struct {
-
-	for _, a := range rs {
-		r.MinTime += a.MinTime
-		r.MaxTime += a.MaxTime
-		r.AvgTime += a.AvgTime
-		r.StdDev += a.StdDev
-		r.SysTime += a.SysTime
-		r.UserTime += a.UserTime
-		r.Rate += a.Rate
-		r.Times = append(r.Times, a.Times...)
-	}
-	l := float64(len(rs))
-	r.MinTime /= l
-	r.MaxTime /= l
-	r.AvgTime /= l
-	r.StdDev /= l
-	r.SysTime /= l
-	r.UserTime /= l
-	r.Rate /= l
-	return r
+		_, err := fmt.Fprintf(s.Writer, strconv.FormatFloat(t/1e9, 'f', 15, 64)+"\n")
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Add another entry that updates the stats
+func (s *CollStats) AddEntry(e Entry) error {
+	switch e.(type) {
+	case CollServerEntry:
+		cse := e.(CollServerEntry)
+		s.round.Update(cse.Time)
+	case CollClientEntry:
+		// what do I want to keep out of the Client Message States
+		// cms.Buckets stores how many were processed at time T
+		// cms.RoundsAfter stores how many rounds delayed it was
+		//
+		// get the average delay (roundsAfter), max and min
+		// get the total number of messages timestamped
+		// get the average number of messages timestamped per second?
+		// get the observed rate of processed messages
+		// avg is how many messages per second, we want how many milliseconds between messages
+		cce := e.(CollClientEntry)
+		avg, _, _, _ := ArrStats(cce.Buckets)
+		observed := avg / 1000
+		observed = 1 / observed
+		s.Rate = observed
+		s.Times = cce.Times
+	case SysEntry:
+		se := e.(SysEntry)
+		s.SysTime = se.SysTime
+		s.UserTime = se.UserTime
+	default:
+		dbg.Fatal("AddEntry did not receive any Coll*Entry.")
+	}
+	return nil
+}
+
+// Average a collection of Stats that better be CollStats !
+func collStatsAverage(stats ...Stats) (Stats, error) {
+	var s *CollStats = new(CollStats)
+	if len(stats) == 0 {
+		return s, errors.New("No stats given to average on CollStats")
+	}
+	first, ok := stats[0].(*CollStats)
+	if !ok {
+		return s, errors.New("Received non CollStats into collStatsAverage")
+	}
+	s.NHosts = first.NHosts
+	s.Depth = first.Depth
+	s.BF = first.BF
+	s.Times = make([]float64, len(first.Times))
+	st := make([]StreamStats, 0, len(stats))
+	for _, b := range stats {
+		switch b.(type) {
+		case *CollStats:
+			a := b.(*CollStats)
+			st = append(st, a.round)
+			s.SysTime += a.SysTime
+			s.UserTime += a.UserTime
+			s.Rate += a.Rate
+			s.Times = append(s.Times, a.Times...)
+		default:
+			return s, errors.New("Average did not receive a CollStats struct")
+		}
+	}
+	s.round = StreamStatsAverage(st...)
+	l := float64(len(stats))
+	s.SysTime /= l
+	s.UserTime /= l
+	s.Rate /= l
+	return s, nil
+}
+
+// Wrapper function to average multiple stats
+// they must be of the same type  !
+func AverageStats(stats ...Stats) (Stats, error) {
+	if len(stats) == 0 {
+		return nil, errors.New("No stats given to average")
+	}
+	switch stats[0].(type) {
+	case *CollStats:
+		return collStatsAverage(stats...)
+	case *ShamirStats:
+		return shamirStatsAverage(stats...)
+	}
```

Comment:
Can't you put that in the interface, like

```
type Stats interface{
...
  CollectStatsAverage([]Stats)(Stats, error)
}
...
func AverageStats(stats []Stats){
...
return stats[0].CollectStats(stats)
}
```

and then define ```CollectStats``` in each implementation of ```Stats```? That way, every new stats that is defined doesn't need his entry in this switch/case part. But then again, I don't know how many of these will exist anyway.